drv/cosmo-hf/src/hf.rs

```diff
@@ -0,0 +1,206 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use drv_hash_api::SHA256_SZ;
+use drv_hf_api::{
+    HfDevSelect, HfError, HfMuxState, HfPersistentData, HfProtectMode,
+};
+use idol_runtime::{
+    LeaseBufReader, LeaseBufWriter, Leased, LenLimit, NotificationHandler,
+    RequestError, R, W,
+};
+use ringbuf::ringbuf_entry_root as ringbuf_entry;
+use userlib::{task_slot, RecvMessage, UnwrapLite};
+
+use crate::{ServerImpl, Trace, PAGE_SIZE_BYTES, SECTOR_SIZE_BYTES};
+
+task_slot!(HASH, hash_driver);
+
+impl idl::InOrderHostFlashImpl for ServerImpl {
+    fn read_id(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<[u8; 20], RequestError<HfError>> {
+        Ok(self.flash_read_id())
+    }
+
+    fn capacity(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<usize, RequestError<HfError>> {
+        todo!()
+    }
+
+    /// Reads the STATUS_1 register from the SPI flash
+    fn read_status(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u8, RequestError<HfError>> {
+        Ok(self.read_flash_status())
+    }
+
+    fn bulk_erase(
+        &mut self,
+        _: &RecvMessage,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if !matches!(protect, HfProtectMode::AllowModificationsToSector0) {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        todo!()
+    }
+
+    fn page_program(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+        data: LenLimit<Leased<R, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr < SECTOR_SIZE_BYTES
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_write(
+            addr,
+            &mut LeaseBufReader::<_, 32>::from(data.into_inner()),
+        )
+        .map_err(|()| RequestError::went_away())
+    }
+
+    /// Erases the 64 KiB sector containing the given address
+    ///
+    /// If this is sector 0, requires `protect` to be
+    /// `HfProtectMode::AllowModificationsToSector0` (otherwise this function
+    /// will return an error).
+    fn read(
+        &mut self,
+        _: &RecvMessage,
+        offset: u32,
+        dest: LenLimit<Leased<W, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        self.flash_read(
+            offset,
+            &mut LeaseBufWriter::<_, 32>::from(dest.into_inner()),
+        )
+        .map_err(|_| RequestError::went_away())
+    }
+
+    fn sector_erase(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr < SECTOR_SIZE_BYTES
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
```

Comment:
nit, take it or leave it: we could maybe have a function like:
```rust
fn check_addr_writable(addr: u32, protect: HfProtectMode) -> Result<(), HfError> {
    if addr < SECTOR_SIZE_BYTES && 
        !matches!(protect, HfProtectMode::AllowModificationsToSector0)
    {
          return Err(HfError::Sector0IsReserved);
    }
    Ok(())
}
```
since we seem to repeat this in a few places. potentially, factoring it out it would mean that a future change adding other checks against `HfProtectMode`s wouldn't have to add the requisite checks in every function that currently checks if sector 0 is writable? 

not a big deal either way though.